cclib/method/bader.py

```diff
@@ -0,0 +1,219 @@
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "Bader's QTAIM charges of {}".format(self.data)
+
+    def __repr__(self):
+        """Return a representation of the object."""
+        return "Bader({})".format(self.data)
+
+    def _check_required_attributes(self):
+        super(Bader, self)._check_required_attributes()
+
+    def calculate(self, indices=None, fupdate=0.05):
+        """Calculate Bader's QTAIM charges using on-grid algorithm proposed by Henkelman group
+           in doi:10.1016/j.commatsci.2005.04.010
+           
+           Cartesian, uniformly spaced grids are assumed for this function.
+           """
+
+        def cartesian_dist(pt1, pt2):
+            """ Small utility function that calculates Euclidian distance between two points
+                pt1 and pt2 are numpy arrays representing a point in Cartesian coordinates """
+            return numpy.sqrt(numpy.einsum("ij,ij->j", pt1 - pt2, pt1 - pt2s))
+
+        # First obtain charge densities on the grid
+        if len(self.data.mocoeffs) == 1:
+            self.chgdensity = electrondensity_spin(
+                self.data, self.volume, [self.data.mocoeffs[0][: self.data.homos[0]]]
+            )
+            self.chgdensity.data *= 2
+        else:
+            self.chgdensity = electrondensity_spin(
+                self.data,
+                self.volume,
+                [
+                    self.data.mocoeffs[0][: self.data.homos[0]],
+                    self.data.mocoeffs[1][: self.data.homos[1]],
+                ],
+            )
+
+        # Assign each grid point to Bader areas
+        self.fragresults = numpy.zeros(self.chgdensity.data.shape, "d")
+        nextIndex = 1
+
+        self.logger.info("Partitioning space into Bader areas.")
+
+        # Exclude the outermost positions
+        for xwalker in range(1, len(self.chgdensity.data) - 1):
+            for ywalker in range(1, len(self.chgdensity.data[0]) - 1):
+                for zwalker in range(1, len(self.chgdensity.data[0][0]) - 1):
+                    (xindex, yindex, zindex) = (xwalker, ywalker, zwalker)
+
+                    if self.fragresults[xindex, yindex, zindex] != 0:
+                        # index has already been assigned for this grid point
+                        continue
+                    else:
+                        listcoord = []
+                        loopcondition = True
+
+                        while loopcondition:
+                            tmp = (
+                                self.chgdensity.data[
+                                    xindex - 1 : xindex + 2,
+                                    yindex - 1 : yindex + 2,
+                                    zindex - 1 : zindex + 2,
+                                ]
+                                - self.chgdensity.data[xindex, yindex, zindex]
+                            )
+                            grad = tmp / griddist
```

Comment:
Can you mark this as being equation (1)? And call it something like `grad_rho_dot_r` or `grad_density_dot_r`?

Reply:
I've just made the comments and renamed the variables. The code looks much more readable this way. Thanks!